docs/adr/2023-11-07_029-streaming-persistence.md

```diff
@@ -0,0 +1,72 @@
+ - An application that only cared about a limited set of message types may be overwhelmed (and slowed down) by the barage of messages on connection
+ - An application that wanted a deeper view into the workings of the hydra protocol would have no access to several of the internal event types 
+
+Additionally, much of the changes that would need to happen to solve any of these would apply equally well to the on-disk persistence the Hydra node currently provides.
+
+-------------------------------------
+
+- Current client API is a stateful WebSocket connection. Client API has its own incrementally persisted state that must be maintained, and includes a monotonically increasing `Seq` ID, but this is based off the events a client observes, not the Hydra host's event log.
+
+- Client must be flexible and ready to handle many different events
+
+- There is no way to simply hand off transactions to the hyrda-node currently, a full connection must be initiated before observed transactions can be applied, and bulky JSON objects can slow down "bulk" operations
+
+- Many applications, like SundaeSwap's Gummiworm Protocol, do not need need the full general API, and benefit from any performance improvement
+
+- The challenge of subscribing to event types is complex to handle, but would be applicable to many parts of Hydra, not just for subscribing to new transactions. It's a good fit for passing stuff off to a message queue (MQTT, Kafka, whatever), probably from a dedicated process (see "Chain Server" below)
+    - Could also just directly use a "compatible" spec like STOMP or MQTT, but that would lock in implementation details
+
+- Previous mock chain using ZeroMQ was [removed][zmq] as part of [#119][#119], due to complexity and feeling of ZeroMQ being unmaintained (primary author no longer contributing, new tag release versions not being published)
+    - This mock chain was used to mock the layer L1 chain, not the L2 ledger
+- Attempt in February 2023 to externalize chainsync server as part of [#230][#230]
+    - Similar to [#119][#119], but the "Chain Server" component in charge of translating Hydra Websocket messages into direct chain, mock chain, or any hypothetical message queue, not necessarily just ZeroMQ
+    - Deemed low priority due to ambiguous use-case at the time. SundaeSwap's Gummiworm Protocol would benefit from the additional control enabled by the Event Server
+
+# Decision
+
+A new abstraction, the EventServer, will be introduced. Each internal hydra event will be sent to the event server, which will be responsible for persisting that event and returning a durable monotonically increasing global event ID. Different implementations of the event server can handle this event differently. Initially, we will implement the following:
+ - MockEventServer, which increments a counter for the ID, and discards the event
```

Comment:
I think the thing this is trying to capture is that sometimes the persistence medium may dictate the id; for example, when writing to a message broker, you may get back a UUID.

Perhaps we don't need to expose *this* id, and could just assign everything a hydra internal ID via a separate counter mechanism, but based on one of the earlier ADRs it sounded like this was a requirement

Reply:
If such an event consumer needs a specific ID, it could derive it from the original event id. I think you are right in saying that this ID would not be exposed and consequently simplify the _fan out_ dynamics here.

Reply:
Well, no, what I'm taking about is when the storage medium itself gives it an ID; for example, you persist the message to kinesis, and kinesis tells you what ID that message had. If we wanted to bubble that up so it could appear in, for example, the websocket APIs. I think the previous ADR we were modelling this off of had this requirement but if you're happy dropping it, I think so are we.